crypto/uri.c

```diff
@@ -0,0 +1,842 @@
+        return 0;
+
+    if (*authority != NULL)
+        return extract_path_abempty(path, pp);
+
+    if (!extract_path_absolute(path, pp)
+        || (*path == NULL && !extract_path_rootless(path, pp))
+        || (*path == NULL && !extract_path_empty(path, pp)))
+        return 0;
+    return 1;
+}
+
+/*
+ * relative-part = "//" authority path-abempty
+ *               / path-absolute
+ *               / path-noscheme
+ *               / path-empty
+ */
+static int extract_relativepart(char **authority, char **path, const char **pp)
+{
+    if (!extract_authority(authority, pp))
+        return 0;
+
+    if (*authority != NULL)
+        return extract_path_abempty(path, pp);
+
+    if (!extract_path_absolute(path, pp)
+        || (*path == NULL && !extract_path_noscheme(path, pp))
+        || (*path == NULL && !extract_path_empty(path, pp)))
+        return 0;
+    return 1;
+}
+
+/*
+ * query         = *( pchar / "/" / "?" )
+ */
+static int extract_query(char **query, const char **pp)
+{
+    const char *p = *pp;
+    size_t off = 0;
+
+    *query = NULL;
+
+    if (!is_valid(p[off])) {
+        URIerr(URI_F_EXTRACT_QUERY, URI_R_INVALID_CHARACTER_IN_URI);
+        return 0;
+    }
+
+    if (p[off] != '?')
+        return 1;
+    p++;
+
+    while(span_pchars(p, &off) || span_chars(p, &off, 0, "/?"))
+        continue;
+
+    /* Check if the loop stopped because of an invalid character */
+    if (!is_valid(p[off])) {
+        URIerr(URI_F_EXTRACT_QUERY, URI_R_INVALID_CHARACTER_IN_URI);
+        return 0;
+    }
+
+    if ((*query = OPENSSL_strndup(p, off)) == NULL) {
+        URIerr(URI_F_EXTRACT_QUERY, ERR_R_MALLOC_FAILURE);
+        return 0;
+    }
+
+    *pp = p + off;
+    return 1;
+}
+
+/*
+ * fragment      = *( pchar / "/" / "?" )
+ */
+static int extract_fragment(char **fragment, const char **pp)
+{
+    const char *p = *pp;
+    size_t off = 0;
+
+    *fragment = NULL;
+
+    if (!is_valid(p[off])) {
+        URIerr(URI_F_EXTRACT_FRAGMENT, URI_R_INVALID_CHARACTER_IN_URI);
+        return 0;
+    }
+
+    if (p[off] != '#')
+        return 1;
+    p++;
+
+    while(span_pchars(p, &off) || span_chars(p, &off, 0, "/?"))
+        continue;
+
+    /* Check if the loop stopped because of an invalid character */
+    if (!is_valid(p[off])) {
+        URIerr(URI_F_EXTRACT_FRAGMENT, URI_R_INVALID_CHARACTER_IN_URI);
+        return 0;
+    }
+
+    if ((*fragment = OPENSSL_strndup(p, off)) == NULL) {
+        URIerr(URI_F_EXTRACT_FRAGMENT, ERR_R_MALLOC_FAILURE);
+        return 0;
+    }
+
+    *pp = p + off;
+    return 1;
+}
+
+/*
+ * URI           = scheme ":" hier-part [ "?" query ] [ "#" fragment ]
+ *
+ * URI-reference = URI / relative-ref
+ *
+ * relative-ref  = relative-part [ "?" query ] [ "#" fragment ]
+ */
+int OPENSSL_decode_uri(const char *uri, char **scheme, char **authority,
+                       char **path, char **query, char **fragment)
+{
+    const char *p = uri;
+
+    OPENSSL_init_crypto(0, NULL);
```

Comment:
It would be good to know. The thing is we are supposed to auto-initialise if we call any functions that need it. These functions don't seem to need it themselves (e.g. no shared global data etc), so if this call was necessary then it suggests some other function call was made that didn't auto-initialise when (possibly) it should have done.